logger.js

```diff
@@ -1,19 +1,204 @@
+    const level = customLogLevel ? customLogLevel(this, err) : useLevel
+
+    if (err || this.err || this.statusCode >= 500) {
+      const error = err || this.err || new Error('failed with status code ' + this.statusCode)
+
+      log[level]({
+        [resKey]: this,
+        [errKey]: error,
+        [responseTimeKey]: responseTime
+      }, errorMessage(error, this))
+      this.alreadyLogged = true
+      return
+    }
+
+    log[level]({
+      [resKey]: this,
+      [finishedKey]: !onlyLogFinishedRequest ? this.writableEnded : true,
+      [responseTimeKey]: responseTime
+    }, successMessage(this))
+    this.alreadyLogged = true
+  }
+
+  function loggingMiddleware (ctx) {
+    let shouldLogSuccess = true
+    const { req, res } = ctx
+
+    // TODO: Is this where we want to set this for koa? (versus ctx or state)
+    // Can we extract request-ids commonly available and/or be compatible with open telemetry/tracing
+    req.id = genReqId(ctx)
+
+    let log = logger.child({ [reqKey]: req })
+
+    if (customProps) {
+      const customPropBindings = (typeof customProps === 'function') ? customProps(ctx) : customProps
+      log = log.child(customPropBindings)
+    }
+
+    ctx.log = req.log = res.log = log
+    ctx.res[startTime] = ctx.res[startTime] || Date.now()
+    ctx.response[startTime] = ctx.response[startTime] || Date.now()
+    ctx[startTime] = ctx[startTime] || Date.now()
+
+    // TODO: do we need to expose the option to filter on the original url?
+    if (autoLogging) {
+      if (autoLoggingIgnorePaths.length) {
+        let url
+        if (autoLoggingGetPath) {
+          url = URL.parse(autoLoggingGetPath(ctx))
+        } else if (ctx.request.url) {
+          url = URL.parse(ctx.request.url)
+        }
+        if (url && url.pathname) {
+          shouldLogSuccess = !autoLoggingIgnorePaths.find(ignorePath => {
+            if (ignorePath instanceof RegExp) {
+              return ignorePath.test(url.pathname)
+            }
+            return ignorePath === url.pathname
+          })
+        }
+      }
+
+      if (shouldLogSuccess) {
+        res.on('finish', onResFinished)
+        if (!onlyLogFinishedRequest) {
+          res.on('close', onResFinished)
+        }
+      }
+
+      res.on('error', onResFinished)
+    }
+  }
+}
+
+function wrapChild (opts, stream) {
+  const prevLogger = opts.logger
+  const prevGenReqId = opts.genReqId
+  let logger = null
+
+  if (prevLogger) {
+    opts.logger = undefined
+    opts.genReqId = undefined
+    logger = prevLogger.child(opts)
+    opts.logger = prevLogger
+    opts.genReqId = prevGenReqId
+  } else {
+    logger = pino(opts, stream)
+  }
+
+  return logger
+}
+
+function reqIdGenFactory (func) {
+  if (typeof func === 'function') return func
+  const maxInt = 2147483647
+  let nextReqId = 0
+  return function genReqId (ctx) {
+    // TODO: decide on order of keys, this matches current behavior of checking req.id
```

Comment:
@jmealo I don't think we need to add support for id on ctx.state. 

The logging middleware should be the first or after an error middleware.

Adding on ctx.req would be the same as other pino adapters.